### PR TITLE
feat: Add `@cerbos/react` package

### PR DESCRIPTION
**Pull request by hasanayan**
This is an initial/draft PR to agree on some of the concepts. It's lacking documentation and testing and some areas that are subject to change.

I am opening this to be able to discuss the what `@cerbos/react` should offer and some other issues on the environment (linter rules etc)

### CODE REVIEW

File: packages/react/src/cerbos-provider.tsx
Comment (hasanayan) [inline packages/react/src/cerbos-provider.tsx:3]: not sure about this rule, maybe importing from peer deps should be allowed?

File: packages/react/package.json
Comment (hasanayan) [inline packages/react/package.json:49]: I am not sure about this, I've added this only for `isEqual` but we might roll our own.

File: packages/react/src/use-check-resource.ts
Comment (hasanayan) [inline packages/react/src/use-check-resource.ts:57]: this is where we need `isEqual`, maybe we should expect the `request` and `options` objects to be memoized and not deal with this?

File: packages/react/src/cerbos-provider.tsx
Comment (haines) [reply packages/react/src/cerbos-provider.tsx:3]: Yeah, agreed, it should be allowed. None of the other packages have peer dependencies so I never ran into this before.

File: packages/react/src/cerbos-provider.tsx
Comment (haines) [inline packages/react/src/cerbos-provider.tsx:13]: I think it would be a bit more ergonomic if the props were

```typescript
client: Client;
principal: Principal;
```

and then we did `value={client.withPrincipal(principal)}`

File: packages/react/src/use-check-resource.ts
Comment (haines) [inline packages/react/src/use-check-resource.ts:19]: Let's just inline these `Omit` types, otherwise we end up adding a whole extra page to the docs for each of them.

File: packages/react/src/use-check-resource.ts
Comment (haines) [reply packages/react/src/use-check-resource.ts:13]: In fact, with sufficient TypeScript magic, we ought to be able to define a single hook so that the other hooks become one-liners

```ts
export function useCheckResource(
  request: Omit<CheckResourceRequest, "principal">,
  options?: RequestOptions
): AsyncResult<CheckResourcesResult> {
  return useRequest("checkResource", request, options);
}
```

**Comment by hasanayan** [comment]
@haines where do you think we should test this package? in `packages/test` with the rest of the packages? We will need to extend the test config with `jsdom` and `@vitejs/plugin-react` but I believe that should be OK. Do you have any other opinion on this?

File: packages/react/src/use-check-resource.ts
Comment (hasanayan) [reply packages/react/src/use-check-resource.ts]: I managed to do this with a `@ts-expect-error` in there 🤮. Luckily none of the ugliness is leaking so I think it's fine unless you can perform a better TS-fu here

**Comment by haines** [comment]
> We will need to extend the test config with `jsdom` and `@vitejs/plugin-react` but I believe that should be OK.

Yeah, that's fine. Although it's maybe less convenient than colocating the tests with the source, the rationale for keeping the tests in a separate package is that it forces us to only test against the publicly-exposed API of the other packages, and it ensures that we don't accidentally ship the tests in the packages we publish to npm.

File: packages/react/src/use-check-resource.ts
Comment (haines) [reply packages/react/src/use-check-resource.ts]: I'll give it a go!

Comment (haines) [reply packages/react/src/use-check-resource.ts]: So, I wasn't able to eliminate it, I realised it's the same problem I have run into several times elsewhere in the SDK. I did make a couple of tweaks while I was trying to do so - see what you think: https://github.com/hasanayan/cerbos-sdk-javascript/pull/1

File: packages/react/src/cerbos-provider.tsx
Comment (haines) [inline packages/react/src/cerbos-provider.tsx:48]: At least one role is required, so the default should be `["anonymous"]` here.

File: packages/react/src/use-cerbos-request.ts
Comment (haines) [inline packages/react/src/use-cerbos-request.ts:97]: Could you run the examples through Prettier please? I usually just paste them into a separate file to do it manually, I haven't found a way to automate it unfortunately.

File: packages/react/src/cerbos-provider.tsx
Comment (haines) [inline packages/react/src/cerbos-provider.tsx:49]: This is a separate prop - I think we can just leave it off for the sake of this example.

```suggestion
```

**Comment by hasanayan** [comment]
@haines I think docs and the actual code is ready for final review, I'm moving on to testing

File: .gitignore
Comment (haines) [inline .gitignore:7]: I try to avoid putting machine/editor-specific ignore patterns in per-project ignore files. This should go in a [global gitignore](https://sebastiandedeyne.com/setting-up-a-global-gitignore-file/), or if you have other projects where you actually do want to commit `.vscode`, then you can put it in `.git/info/exclude` which is specific to your copy of this repo.

**Review by haines — COMMENTED** [review]
Code looks great. I've made some minor suggestions for the docs on https://github.com/hasanayan/cerbos-sdk-javascript/pull/2. We'll also need to add a changelog but we should do that after #18 lands because I changed the format 🙈

**Comment by hasanayan** [comment]
@haines added tests with a placeholder test for testing request cancellation, I'll wait until you merge your PR and I'll add the test.

I also added changelog.yml file but I am not sure it's correct. Btw I don't know what should be our initial version, an alpha or beta maybe? For now it's set as 0.0.1 in package.json

Also, apparently my brain is still in weekend, I somehow thought your docs PR was actually a direct commit (you don't even have access, I know ^^)  and I force pushed and removed it so I manually took those changes in 🤦🏻 😄

File: packages/react/changelog.yaml
Comment (haines) [inline packages/react/changelog.yaml:11]: `initialCommit` is maybe not the right name here in retrospect. It's the last commit before the package was introduced (i.e. where to start diffing from for the first release). I'll avoid merging anything else before this one now, so that we don't have to keep updating this.

```suggestion
initialCommit: 68337848cb3f987627da7381653d82c6d4e368a5
```

File: packages/react/changelog.yaml
Comment (haines) [inline packages/react/changelog.yaml:10]: Let's just drop this for now, we won't be dropping support for any Node.js versions until 18 is EOL at the end of April next year.

```suggestion
```

File: packages/react/package.json
Comment (haines) [inline packages/react/package.json:3]: Let's go with 0.1.0 for the initial release (I don't think we need to mark it as alpha/beta - we only did that for the embedded client because we didn't yet have full support for all possible policies in the WASM compiler).

However, with the new automation, we will do a minor version bump (as specified in the changelog) to the number we put here before actually releasing, so for now it should probably be:

```suggestion
  "version": "0.0.1",
```

File: packages/react/package.json
Comment (haines) [inline packages/react/package.json:46]:
```suggestion
    "React",
    "hooks"
```

File: private/test/src/react/use-check-resource.test.tsx
Comment (haines) [inline private/test/src/react/use-check-resource.test.tsx:1]: Maybe just `hooks.test.tsx` since we have them all in here?

File: private/test/src/react/use-check-resource.test.tsx
Comment (haines) [inline private/test/src/react/use-check-resource.test.tsx:216]: I think you can just `import { setTimeout } from "timers/promises"` (unless that doesn't work with the fake timers?).

**Review by haines — COMMENTED** [review]
This is looking great!

File: private/test/src/react/use-check-resource.test.tsx
Comment (hasanayan) [reply private/test/src/react/use-check-resource.test.tsx:216]: yep it doesn't work with the fake timers. But I didn't know about this, thanks!

**Comment by hasanayan** [comment]
@haines all done but; I broke my back trying to get React to complain about setState call after unmounting during tests (without `if (!abortController.signal.aborted)` check in the catch block) so that I could test that happens. Unfortunately I could succeed. But I am testing that we always pass an AbortSignal with the requests and the signal is aborted on unmount.

File: packages/react/src/use-cerbos-request.ts
Comment (haines) [inline packages/react/src/use-cerbos-request.ts:63]: I think the `then` gets enqueued after the promise gets resolved, so in the pathological case there might be a possibility that an abort happens in between the `fetch` resolving and this being called?

To be defensive, perhaps we should keep the `if (!aborted)` logic we had before.

File: private/test/src/react/hooks.test.tsx
Comment (haines) [inline private/test/src/react/hooks.test.tsx:231]: Minor but here and below you could probably avoid the non-null assertion because the test will still fail if it is undefined:

```suggestion
      expect(abortSignal?.aborted).toBe(false);
```

**Comment by haines** [comment]
> I broke my back trying to get React to complain about setState call after unmounting during tests (without `if (!abortController.signal.aborted)` check in the catch block) so that I could test that happens. Unfortunately I could succeed. But I am testing that we always pass an AbortSignal with the requests and the signal is aborted on unmount.

Yeah, sounds painful - I think we have good test coverage of the abort behaviour between the tests I wrote for the client and the one you have added here, so I'm happy enough with that!

File: private/test/src/react/hooks.test.tsx
Comment (hasanayan) [reply private/test/src/react/hooks.test.tsx:231]: I thought the same but `no-unsafe-assignment` rule is not happy about that (no configuration options). I chose this one to suppress

Comment (haines) [reply private/test/src/react/hooks.test.tsx:231]: Ah, ok... weird!

**Review by haines — APPROVED** [review]
Awesome!

**Pull request merged**
